## **Урок 1. Случайные события. Условная вероятность. Формула Байеса. Независимые испытания**

#### Задача 1. Из колоды в 52 карты извлекаются случайным образом 4 карты.
#### a) Найти вероятность того, что все карты – крести. 
#### б) Найти вероятность, что среди 4-х карт окажется хотя бы один туз.

Начнем с того, что создадим функцию для нахождения количества сочетаний.

In [63]:
from math import factorial

In [64]:
def combinations(n, k):
    return int(factorial(n) / (factorial(k) * factorial(n - k)))

#### a) Найти вероятность того, что все карты – крести.
Вероятность находится по формуле:

$P = \frac{число \;исходов,\; благоприятствующих\;  событию}{общее\; число \;исходов }$

число исходов, благоприятствующих событию – это количество сочетаний 4 карт из 52, где будет 4 крестовых карты (так как в задаче сказано, что ВСЕ карты крести), C1 

общее число исходов – это все возможные сочетания 4 карт из 52, С2

Получается, вероятность будет находиться по формуле:

$P = \frac{C1}{C2}$

Найдем C1, C2 и подставим в формулу.

In [65]:
#Найдем С1

#Получается, вытаскиваются 4 крестовых карты и 0 карт других мастей. 
#В колоде 52 карты по 13 карт каждой масти.

#Найдем число сочетаний из 13 по 4 карты

С1_1 = combinations(13, 4)
С1_1

715

In [66]:
#Осталось 39 некрестовых карт

#Найдем число сочетаний из 39 по 0 карт

С1_2 = combinations(39, 0)
С1_2

1

In [67]:
#Найдем С1

C1 = С1_1 * С1_2

C1

715

In [68]:
#Теперь найдем С2
#Это общее число исходов – это все возможные сочетания 4 карт из 52

C2 = combinations(52, 4)
C2

270725

In [69]:
P = C1 / С2
P

0.0026410564225690276

#### Ответ: вероятность того, что все 4 карты – крести составляет 0.0026410564225690276 или примерно 0,26%

#### б) Найти вероятность, что среди 4-х карт окажется хотя бы один туз.

То есть из 4-х карт будет или 1, или 2, или 3, или 4 туза.

Аналогично, вероятность находится по формуле:

$P = \frac{число \;исходов,\; благоприятствующих\;  событию}{общее\; число \;исходов }$

число исходов, благоприятствующих событию – это количество сочетаний 4 карт из 52, где будет 1, 2, 3 или 4 туза, С3 

общее число исходов – это все возможные сочетания 4 карт из 52, С2. Мы его уже находили в прошлый раз С2 = 270725

Получается вероятность будет находиться по формуле:

$P = \frac{C3}{C2}$

Найдем C3 и подставим в формулу.

Найдем С3:

$C3 = C3_1 + C3_2 + C3_3 + C3_4$

C3_1, C3_2, C3_3, C3_4 - это количество комбинаций с 1, 2, 3, 4 тузами соответственно.

In [70]:
#Найдем С3_1, то есть рассмотрим ситуацию, когда у нас 1 туз и 3 не туза.
#Найдем число сочетаний из 4 тузов по 1-му

C3_1_1 = combinations(4, 1)
C3_1_1 

4

In [71]:
# 48 оставшихся карт - это не тузы.
# Надем из них сочетание по 3.

C3_1_2 = combinations(48, 3)
C3_1_2 

17296

In [72]:
C3_1 = C3_1_2 * C3_1_1 
C3_1

69184

In [73]:
#Оформим получившиеся расчеты в виде функции, чтобы не повторять код.
#n - количество тузов

def ace_combinations(n):
    return combinations(4, n) * combinations(48, 4-n)

C3_1_test = ace_combinations(1)
C3_1_test

69184

In [74]:
C3_2 = ace_combinations(2)
C3_2

6768

In [75]:
C3_3 = ace_combinations(3)
C3_3

192

In [76]:
C3_4 = ace_combinations(4)
C3_4 

1

In [77]:
C3  = C3_1 + C3_2 + C3_3 + C3_4 
C3

76145

In [78]:
P_1 = C3 / C2
P_1

0.2812632745405855

#### Ответ:  вероятность, что среди 4-х карт окажется хотя бы один туз составляет 0.2812632745405855 или примерно 28,13%

#### Задача 2. На входной двери подъезда установлен кодовый замок, содержащий десять кнопок с цифрами от 0 до 9.
#### Код содержит три цифры, которые нужно нажать одновременно.
#### Какова вероятность того, что человек, не знающий код, откроет дверь с первой попытки?

Получается, человек нажмет одновременно три кнопки. Так как, нажав первую, он на нее второй раз нажать не сможет, 
то эти три события являются зависимыми. 

Зависимыми событиями называются, когда появление одного влияет на появление другого.

Вероятность появления двух зависимых событий

$P(A*B) = P(A)* P(B|A)$

Рассчитаем данную вероятность для трех событий задачи.

In [79]:
# Вероятность нажатия правильно первой кнопки
A = 1 / 10

#Вероятность нажатия правильно второй кнопки
B = 1 / 9

#Вероятность нажатия правильно третьей кнопки
C = 1 / 8

A, B, C

(0.1, 0.1111111111111111, 0.125)

In [80]:
P = A * B * C
P

0.001388888888888889

#### Ответ:  вероятность,  что человек, не знающий код, откроет дверь с первой попытки - 0.001388888888888889 или  примерно 0,14%

#### Задача 3. В ящике имеется 15 деталей, из которых 9 окрашены. 
#### Рабочий случайным образом извлекает 3 детали. 
#### Какова вероятность того, что все извлеченные детали окрашены?

Когда рабочий вытащит одну окрашенную деталь, их количество уменьшится. Значит, последующие случаи извлечения окрашенной 
детали - это зависимые события.

Зависимыми событиями называются, когда появление одного влияет на появление другого.

Вероятность появления двух зависимых событий

$P(A*B) = P(A)* P(B|A)$

Рассчитаем данную вероятность для трех событий задачи.

In [81]:
#Вероятность извлечения первой окрашенной детали
A = 9 / 15

#Вероятность извлечения второй окрашенной детали
B = 8 / 14

#Вероятность извлечения третьей окрашенной детали
C = 7 / 13

A, B, C

(0.6, 0.5714285714285714, 0.5384615384615384)

In [82]:
P = A * B * C
P

0.18461538461538457

#### Ответ:  вероятность, что все извлеченные детали окрашены - 0.18461538461538457 или примерно 18,5%

#### Задача 4. В лотерее 100 билетов. Из них 2 выигрышных.
#### Какова вероятность того, что 2 приобретенных билета окажутся выигрышными?

Также, как в предыдущих задачах, эти два события (приобретение выигрышных билетов) являются зависимыми.

Зависимыми событиями называются, когда появление одного влияет на появление другого.

Вероятность появления двух зависимых событий

$P(A*B) = P(A)* P(B|A)$

Рассчитаем данную вероятность для двух событий задачи.

In [83]:
#Вероятность приобретения первого выигрышного билета
A = 2 / 100

#Вероятность приобретения второго выигрышного билета
B = 1 / 99

A, B

(0.02, 0.010101010101010102)

In [84]:
P = A * B
P

0.00020202020202020205

#### Ответ:  вероятность, что 2 приобретенных билета окажутся выигрышными 0.00020202020202020205 или примерно 0,02%